### PR TITLE
docs: contract storage design doc

### LINKED ISSUE #293: dev: explore an accumulator based pattern to apply storage updates
> we could track modified storage addresses in a dict, and sload/sstore would read/write to this dict, and if the value is 0 write/read to storage instead. We would track all modified storage addresses in an array and then loop through when the current context exits.

> Might be more interesting gas-wise as if a subcontext reverts, we divide the read/write syscalls amount by two

Exploring this feature can be interesting as it is way less expansive to handle subcontext reverts. However, this might introduce overhead costs for OK cases

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #359 Resolves #293 

## What is the new behavior?

- Improvements in the Contract Storage design doc
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: docs/general/model/contract_storage.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
+# Kakarot Storage
+
+## Storage in Ethereum
+
+The top-level data structure that holds information about the state of the
+Ethereum blockchain is called the _world state_, and is a mapping of Ethereum
+addresses (160-bit values) to accounts. Each Ethereum address represents an
+account composed by a _nonce_, an _ether balance_, a _storage_, and a _code_. We
+make the distinction between EOA (Externally Owned Accounts) that have no code
+and an empty storage, and contracts that can have code and storage.
```
Comment (Eikix) [inline docs/general/model/contract_storage.md:10]: Here we could add an image from https://takenobu-hs.github.io/downloads/ethereum_evm_illustrated.pdf and add the source

Smth like 
image

File: docs/general/model/contract_storage.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+The behavior for the SLOAD opcode is very complementary to the SSTORE opcode.
+When Geth executes the SLOAD opcode, it does the following:
+
+- It starts by doing a check on `dirtyStorage` to see if it contains a value for
+  the queried key, and returns it if so.
+- Otherwise, it retrieves the value from the committed account storage trie.
+
+Since one transaction can access a storage slot multiple times, we must ensure
+that the result returned is the most recent value. This is why Geth first checks
+`dirtyStorage`, which is the most up-to-date state of the storage.
```
Comment (Eikix) [inline docs/general/model/contract_storage.md:39]: I think this section deserves a smol mermaid diagram

File: docs/general/model/contract_storage.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  in each Kakarot contract. This is not ideal from a security perspective.
+- It moves away from the traditional EVM design, in which execution clients
+  store account states in a common database backend.
+
+Therefore, we will not use this design in SSJ. We will instead use the second
+design presented thereafter.
+
+### A shared storage space for all Kakarot Contracts
+
+The second approach is to have a single storage space for all Kakarot contracts.
```
Comment (Eikix) [inline docs/general/model/contract_storage.md:90]:
```suggestion
The second approach is to have a single storage space for all Kakarot contract accounts.
```

File: docs/general/model/contract_storage.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+The SSTORE and SLOAD opcodes are implemented to first read and write to the
+`Journal` instead of directly writing to the KakarotCore contract's storage.
+
+Using the `storage_read_syscall` and `storage_write_syscall` syscalls, we can
+arbitrarily read and write to a contract's storage. Therefore, we will be able
+to simply implement the SSTORE and SLOAD opcodes as follows:
+
+```rust
+  // SSTORE
+  let storage_address = poseidon_hash(evm_address, storage_slot);
```
Comment (Eikix) [inline docs/general/model/contract_storage.md:180]: is storage_slot in this situation a u256?

File: docs/general/model/contract_storage.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+A contract’s storage on Starknet is a persistent storage space where you can
+read, write, modify, and persist data. The storage is a map with $2^{251}$
+slots, where each slot is a felt which is initialized to 0.
+
+This new model doesn't expose read and write methods on Kakarot contracts.
+Instead of having $n$ contracts with `write_storage` and `read_storage`
+entrypoints, the only way to update the storage of a Kakarot contract is now
+through executing SLOAD / SSTORE internally to KakarotCore.
+
+```mermaid
```
Comment (Eikix) [inline docs/general/model/contract_storage.md:107]: should we add the accumulator as a participant between interpreter and contractstate?

**Review by Eikix — COMMENTED** [review]
pretty awesome! looking forward to seeing it in action

File: docs/general/model/contract_storage.md (hunk 4 above)
Comment (enitrat) [reply docs/general/model/contract_storage.md:180]: Yes, it's the key popped from the stack which is a 32-bytes value

**Comment by enitrat** [comment]
Applied suggestions @Eikix

**Review by Eikix — APPROVED** [review]
lgtm 🔥🔥🔥

**Pull request merged**